scio-elasticsearch/src/main/java/com/spotify/scio/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,265 @@
+      return new Bound<T>().withClusterName(clusterName);
+    }
+
+    /**
+     * Returns a transform for writing to the Elasticsearch cluster for a given servers.
+     *
+     * @param servers endpoints for the Elasticsearch cluster
+     */
+    public static<T> Bound<T> withServers(InetSocketAddress[] servers) {
+      return new Bound<T>().withServers(servers);
+    }
+
+    /**
+     * Returns a transform for writing to Elasticsearch cluster by providing
+     * slight delay specified by flushInterval.
+     *
+     * @param flushInterval delay applied to buffer elements. Defaulted to 1 seconds.
+     */
+    public static<T> Bound withFlushInterval(Duration flushInterval) {
+      return new Bound<T>().withFlushInterval(flushInterval);
+    }
+
+    /**
+     * Returns a transform for writing to Elasticsearch cluster.
+     *
+     * @param function creates IndexRequest required by Elasticsearch client
+     */
+    public static<T> Bound withFunction(SerializableFunction function) {
+      return new Bound<T>().withFunction(function);
+    }
+
+    /**
+     * Returns a transform for writing to Elasticsearch cluster.
+     * Note: Recommended to set this number as number of workers in your pipeline.
+     *
+     * @param numOfShard to construct a batch to bulk write to Elasticsearch.
+     */
+    public static<T> Bound withNumOfShard(Long numOfShard) {
+      return new Bound<>().withNumOfShard(numOfShard);
+    }
+
+    public static class Bound<T> extends PTransform<PCollection<T>, PDone> {
+      private final String clusterName;
+      private final InetSocketAddress[] servers;
+      private final Duration flushInterval;
+      private final SerializableFunction<T, IndexRequest> function;
+      private final Long numOfShard;
+
+      private Bound(final String clusterName,
+                    final InetSocketAddress[] servers,
+                    final Duration flushInterval,
+                    final SerializableFunction<T, IndexRequest> function,
+                    final Long numOfShard) {
+        this.clusterName = clusterName;
+        this.servers = servers;
+        this.flushInterval = flushInterval == null? Duration.ofSeconds(1L): flushInterval;
+        this.function = function;
+        this.numOfShard = numOfShard;
+      }
+
+      Bound() {
+        this(null, null, null, null, null);
+      }
+
+      public Bound<T> withClusterName(String clusterName) {
+        return new Bound<>(clusterName, servers, flushInterval, function, numOfShard);
+      }
+
+      public Bound<T> withServers(InetSocketAddress[] servers) {
+        return new Bound<>(clusterName, servers, flushInterval, function, numOfShard);
+      }
+
+      public Bound<T> withFlushInterval(Duration flushInterval) {
+        return new Bound<>(clusterName, servers, flushInterval, function, numOfShard);
+      }
+
+      public Bound<T> withFunction(SerializableFunction<T, IndexRequest> function) {
```

Comment:
We need to provide a custom coder in that case. Its better to abstract away application to to that transformation.